### PR TITLE
Allow to double Throwable, or an interface that extends Throwable

### PR DESCRIPTION
**Pull request by ciaranmcnulty**
The use case is explained in #410 

When doubling something like the following:

```
interface VendorProblem extends Throwable {}
```

 Prophecy was trying to generate something like:

```
class DoubleOfFoo extends stdClass implements MyTraversable
```

This is not allowed by PHP. To fix, there is a ClassPatch that sets the class being extended to Exception instead of stdClass

### CODE REVIEW

File: src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        return false;
+    }
+
+    /**
+     * @param string $type
+     * @return bool
+     */
+    private function isThrowable($type)
```
Comment (stof) [inline src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php]: Do we really need that many private methods ?

File: src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        $this->setParentClassToException($node);
+    }
+
+    /**
+     * @param ClassNode $node
+     */
+    private function checkItCanBeDoubled(ClassNode $node)
+    {
+        $className = $node->getParentClass();
+        if ($className != 'stdClass') {
```
Comment (stof) [inline src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php]: should use strict comparison

File: src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+     * @return void
+     */
+    public function apply(ClassNode $node)
+    {
+        $this->checkItCanBeDoubled($node);
+        $this->setParentClassToException($node);
+    }
+
+    /**
+     * @param ClassNode $node
```
Comment (stof) [inline src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php]: useless phpdoc. It duplicates the typehint

Comment (ciaranmcnulty) [reply src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php]: Wasn't sure what the project policy was :-) will strip a few out in other places too then

File: src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php (hunk 1 above)
Comment (ciaranmcnulty) [reply src/Prophecy/Doubler/ClassPatch/ThrowablePatch.php]: I tend to write this way but if you don't like it I can combine a few of them? Maybe it'll look better without the unnecessary docblocks

**Comment by stof** [comment]
Tests are failing on PHP 5.x. The specs need to be skipped on 5.x when they expect the Throwable interface to exist

File: fixtures/ThrowableInterface.php [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace Fixtures\Prophecy;
+
+interface ThrowableInterface extends \Throwable
+{
+}
```
Comment (dkarlovi) [inline fixtures/ThrowableInterface.php]: Newline missing.

File: spec/Prophecy/Doubler/ClassPatch/ThrowablePatchSpec.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        if (\PHP_VERSION_ID < 70000) {
+            throw new SkippingException('Throwable is not defined in PHP 5');
+        }
+
+        $node->getParentClass()->willReturn('ArrayObject');
+
+        $this->shouldThrow('Prophecy\Exception\Doubler\ClassCreatorException')->duringApply($node);
+    }
+}
```
Comment (dkarlovi) [inline spec/Prophecy/Doubler/ClassPatch/ThrowablePatchSpec.php]: Newline missing.

**Comment by dkarlovi** [comment]
Hopefully, this allows https://github.com/symfony/symfony/pull/26702 to be re-applied. /cc @dunglas @nicolas-grekas

**Comment by ciaranmcnulty** [comment]
@stof can you take another look please?

**Pull request merged**